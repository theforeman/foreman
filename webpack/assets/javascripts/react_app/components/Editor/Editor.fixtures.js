import { noop } from '../../common/helpers';

export const editor = {
  name: 'editor',
  diffViewType: 'split',
  value: 'value',
  previewValue: 'preview',
  renderedEditorValue: 'renderedEditorValue',
  hosts: [
    { id: '1', name: 'host1' },
    { id: '2', name: 'host2' },
  ],
  filteredHosts: [],
  previewResult: 'previewResult',
  mode: 'Ruby',
  theme: 'Monokai',
  keyBinding: 'Default',
  selectedView: 'input',
  editorName: 'editor',
  isMaximized: false,
  isMasked: false,
  isRendering: false,
  isLoading: false,
  isFetchingHosts: false,
  isSearchingHosts: false,
  readOnly: false,
  showError: true,
  errorText: '',
  selectedHost: { id: '', name: '' },
  isSelectOpen: false,
  searchQuery: '',
  template: '',
};

export const dropdowns = {
  modes: ['Text', 'Json', 'Ruby', 'Html_ruby', 'Sh', 'Xml', 'Yaml'],
  keyBindings: ['Default', 'Emacs', 'Vim'],
  themes: ['Github', 'Monokai'],
};

export const showBooleans = {
  showPreview: true,
  showHostSelector: true,
  showImport: true,
  showHide: true,
};

export const editorOptions = {
  ...editor,
  ...dropdowns,
  data: {
    ...showBooleans,
    name: 'editor',
    title: 'title',
    template: '<? />',
    options: dropdowns,
  },
  initializeEditor: noop,
  importFile: noop,
  revertChanges: noop,
  previewTemplate: noop,
  toggleModal: noop,
  changeDiffViewType: noop,
  changeEditorValue: noop,
  dismissErrorToast: noop,
  changeTab: noop,
  toggleMaskValue: noop,
  changeSetting: noop,
  toggleRenderView: noop,
  onHostSearch: noop,
  onHostSelectToggle: noop,
  onSearchClear: noop,
  fetchAndPreview: noop,
  template: '<? />',
  renderPath: '/render/path',
};

export const serverRenderResponse = {
  data: ['< renderedData />'],
};

export const hosts = [
  { id: '1', name: 'host1' },
  { id: '2', name: 'host2' },
];
