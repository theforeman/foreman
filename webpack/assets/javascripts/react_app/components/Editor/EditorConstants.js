export const EDITOR_INITIALIZE = 'EDITOR_INITIALIZE';
export const EDITOR_IMPORT_FILE = 'EDITOR_IMPORT_FILE';
export const EDITOR_REVERT_CHANGES = 'EDITOR_REVERT_CHANGES';
export const EDITOR_EXEC_PREVIEW = 'EDITOR_EXEC_PREVIEW';
export const EDITOR_MODAL_TOGGLE = 'EDITOR_MODAL_TOGGLE';
export const EDITOR_CHANGE_DIFF_VIEW = 'EDITOR_CHANGE_DIFF_VIEW';
export const EDITOR_CHANGE_VALUE = 'EDITOR_CHANGE_VALUE';
export const EDITOR_SHOW_ERROR = 'EDITOR_SHOW_ERROR';
export const EDITOR_DISMISS_ERROR = 'EDITOR_DISMISS_ERROR';
export const EDITOR_CHANGE_TAB = 'EDITOR_CHANGE_TAB';
export const EDITOR_TOGGLE_MASK = 'EDITOR_TOGGLE_MASK';
export const EDITOR_CHANGE_SETTING = 'EDITOR_CHANGE_SETTING';
export const EDITOR_TOGGLE_RENDER_VIEW = 'EDITOR_TOGGLE_RENDER_VIEW';
export const EDITOR_SHOW_LOADING = 'EDITOR_SHOW_LOADING';
export const EDITOR_HIDE_LOADING = 'EDITOR_HIDE_LOADING';
export const EDITOR_FETCH_HOST_PENDING = 'EDITOR_FETCH_HOST_PENDING';
export const EDITOR_FETCH_HOST_RESOLVED = 'EDITOR_FETCH_HOST_RESOLVED';
export const EDITOR_HOST_SELECT_TOGGLE = 'EDITOR_HOST_SELECT_TOGGLE';
export const EDITOR_HOST_SELECT_CLEAR = 'EDITOR_HOST_SELECT_CLEAR';
export const EDITOR_HOST_SELECT_RESET = 'EDITOR_HOST_SELECT_RESET';
export const EDITOR_HOST_INITIAL_FETCH = 'EDITOR_HOST_INITIAL_FETCH';

export const EDITOR_HOSTS_URL = '/hosts/preview_host_collection.json';
export const EDITOR_HOST_ARR = 'hosts';
export const EDITOR_HOST_FILTERED_ARR = 'filteredHosts';
export const EDITOR_KEYBINDINGS = ['Default', 'Emacs', 'Vim'];
export const EDITOR_THEMES = ['Github', 'Monokai'];
export const EDITOR_MODES = [
  'Text',
  'Json',
  'Ruby',
  'Html_ruby',
  'Sh',
  'Xml',
  'Yaml',
];
