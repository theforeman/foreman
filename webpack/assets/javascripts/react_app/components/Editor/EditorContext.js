import { createContext } from 'react';

export const EditorContext = createContext(null);
