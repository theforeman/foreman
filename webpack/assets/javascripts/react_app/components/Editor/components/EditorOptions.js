/* eslint-disable no-alert */
import React from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Icon,
  OverlayTrigger,
  FormControl,
  Tooltip,
} from 'patternfly-react';

import { translate as __ } from '../../../common/I18n';
import { bindMethods } from '../../../common/helpers';
import DiffRadioButtons from '../../DiffView/DiffRadioButtons';
import EditorSettings from './EditorSettings';

class EditorOptions extends React.Component {
  constructor(props) {
    super(props);
    bindMethods(this, ['fileDialog']);
    this.fileInput = React.createRef();
  }

  fileDialog() {
    this.fileInput.click();
  }

  render() {
    const {
      changeDiffViewType,
      changeSetting,
      changeTab,
      diffViewType,
      importFile,
      isDiff,
      isMasked,
      keyBinding,
      keyBindings,
      mode,
      modes,
      revertChanges,
      selectedView,
      showHide,
      showImport,
      template,
      theme,
      themes,
      toggleMaskValue,
      toggleModal,
    } = this.props;

    return (
      <div id="editor-dropdowns">
        {selectedView === 'diff' && (
          <DiffRadioButtons
            stateView={diffViewType}
            changeState={viewType => changeDiffViewType(viewType)}
          />
        )}

        <h4 id="divider">|</h4>
        {showHide && (
          <OverlayTrigger
            delayShow={500}
            overlay={<Tooltip id="mask-tooltip">{__('Hide Content')}</Tooltip>}
            placement="top"
            trigger={['hover']}
          >
            <Button
              disabled={selectedView !== 'input'}
              className="editor-button"
              id="hide-btn"
              onClick={() => toggleMaskValue(isMasked)}
              bsStyle="link"
            >
              <Icon size="lg" type="fa" name={isMasked ? 'eye' : 'eye-slash'} />
            </Button>
          </OverlayTrigger>
        )}
        {isDiff ? ( // fixing tooltip showing sometimes for disabled icon
          <OverlayTrigger
            delayShow={500}
            overlay={
              <Tooltip id="revert-tooltip">
                {__('Revert Local Changes')}
              </Tooltip>
            }
            placement="top"
            trigger={['hover']}
          >
            <Button
              className="editor-button"
              id="undo-btn"
              onClick={() => {
                if (
                  window.confirm(
                    'Are you sure you would like to revert all changes?'
                  )
                ) {
                  revertChanges(template);
                  if (selectedView !== 'input') changeTab('input');
                }
              }}
              bsStyle="link"
            >
              <Icon size="2x" type="pf" name="restart" />
            </Button>
          </OverlayTrigger>
        ) : (
          <Button
            disabled
            className="editor-button"
            id="undo-btn"
            bsStyle="link"
          >
            <Icon size="2x" type="pf" name="restart" />
          </Button>
        )}
        {showImport && (
          <OverlayTrigger
            delayShow={500}
            overlay={<Tooltip id="import-tooltip">{__('Import File')}</Tooltip>}
            placement="top"
            trigger={['hover']}
          >
            <Button
              disabled={selectedView !== 'input'}
              className="import-button"
              id="import-btn"
              bsStyle="link"
              onClick={() => this.fileDialog()}
            >
              <Icon size="lg" type="pf" name="folder-open" />
              <FormControl
                inputRef={ref => {
                  this.fileInput = ref;
                }}
                className="hidden"
                type="file"
                onChange={importFile}
              />
            </Button>
          </OverlayTrigger>
        )}
        <EditorSettings
          changeSetting={changeSetting}
          selectedView={selectedView}
          modes={modes}
          mode={mode}
          keyBindings={keyBindings}
          keyBinding={keyBinding}
          theme={theme}
          themes={themes}
        />
        <OverlayTrigger
          delayShow={500}
          overlay={<Tooltip id="fullscreen-tooltip">{__('Maximize')}</Tooltip>}
          placement="top"
          trigger={['hover']}
        >
          <Button
            className="editor-button"
            id="fullscreen-btn"
            onClick={toggleModal}
            bsStyle="link"
          >
            <Icon size="lg" type="fa" name="arrows-alt" />
          </Button>
        </OverlayTrigger>
      </div>
    );
  }
}

EditorOptions.propTypes = {
  changeDiffViewType: PropTypes.func.isRequired,
  changeSetting: PropTypes.func.isRequired,
  changeTab: PropTypes.func.isRequired,
  diffViewType: PropTypes.string.isRequired,
  importFile: PropTypes.func.isRequired,
  isDiff: PropTypes.bool.isRequired,
  isMasked: PropTypes.bool.isRequired,
  keyBinding: PropTypes.string.isRequired,
  keyBindings: PropTypes.array.isRequired,
  mode: PropTypes.string.isRequired,
  modes: PropTypes.array.isRequired,
  revertChanges: PropTypes.func.isRequired,
  selectedView: PropTypes.string.isRequired,
  showHide: PropTypes.bool,
  showImport: PropTypes.bool.isRequired,
  template: PropTypes.string,
  theme: PropTypes.string.isRequired,
  themes: PropTypes.array.isRequired,
  toggleMaskValue: PropTypes.func.isRequired,
  toggleModal: PropTypes.func.isRequired,
};

EditorOptions.defaultProps = {
  showHide: false,
  template: '',
};

export default EditorOptions;
