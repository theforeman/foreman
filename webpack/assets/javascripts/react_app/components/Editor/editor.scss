.ace_content {
  font-family:
    'SFMono-Regular',
    Consolas,
    'Liberation Mono',
    Menlo,
    Courier,
    monospace;
}

.mask-editor {
  .ace_content {
    font-family: 'AllBullets', monospace;
  }
}

.popover-content {
  padding: 2px;

  .dropdown {
    text-align: center;
    margin-bottom: 5px;

    button {
      color: #1e1d24;
    }

    #settings-dropdown {
      min-width: unset;
    }
  }
}

.editor-modal {
  .modal-content {
    height: -webkit-fill-available;
  }

  .modal-header {
    #diff-radio-buttons {
      margin-bottom: unset;
      margin-right: 15px;

      button {
        outline: none;
      }
    }

    #editor-modal-h4 {
      margin-top: 4px;
      margin-bottom: 0;
      float: left;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      max-width: 700px;
    }
  }

  .modal-dialog {
    padding: 15px;
    width: auto !important;

    .input.modal-body {
      background-color: #363636;
    }

    .modal-body {
      padding: 0;
      overflow: auto;
      height: 100%;

      .ace_editor {
        border: none !important;
      }

      #diff-table {
        max-height: 80vh;

        div {
          border: none;
        }
      }
    }

    .input.modal-header.monokai,
    .preview.modal-header.monokai {
      background-color: #363636;

      h4 {
        color: #fff;
      }

      .pficon-close {
        color: #fff;
      }
    }
  }
}

.navbar-editor {
  padding: 2px 2px 0 2px !important;
  margin: 10px 0 0 !important;
  height: 40px;
  min-height: unset !important;

  .nav-tabs-pf-secondary {
    float: left;
    background: unset !important;
    border-bottom: unset;
    display: flex;

    a {
      background: unset !important;
    }

    .dropdown-menu > li > a:hover {
      background-color: #08c !important;
    }

    #editor-select-container {
      width: 260px;
      margin-top: 4px;

      .select-container-pf {
        border-radius: 2px;
      }
    }

    #outdated-preview-alert {
      display: flex;
      align-items: center;

      .alert {
        padding: 0;
        display: flex;
        margin: 0 5px;
        height: 24px;
        align-items: center;
        border-radius: 2px;

        .pficon-warning-triangle-o {
          position: inherit;
          font-size: 15px;
          top: 0;
          left: 0;
          margin: 0 5px;
        }
      }
    }

    #preview-spinner {
      align-items: center;
      display: flex;
      margin-left: 10px;
    }
  }

  .btn-group {
    padding: 4px;
  }

  .btn[disabled] {
    cursor: unset;
  }

  #editor-dropdowns {
    float: right;

    #diff-radio-buttons {
      float: left;
      margin-right: 5px;

      button {
        outline: none;
      }
    }

    #divider {
      float: left;
      font-size: 18px;
      margin-top: 6px;
      font-weight: 200;
      color: #c5c5c5;
      cursor: default;
      user-select: none;
    }

    button:enabled :not(#diff-radio-buttons) {
      color: #1e1d24;
    }

    button:hover:enabled :not(#diff-radio-buttons) {
      color: rgb(75, 79, 94) !important;
    }
  }

  .preview_type_ahead {
    float: right;
    padding: 4px;

    input {
      width: 275px;
    }

    .rbt-aux {
      right: 0;
    }
  }
}

#hide-btn {
  width: 40px !important;
  font-size: 15px;
  margin-top: -4px;
}

#fullscreen-btn {
  width: 40px !important;
  font-size: 14px;
  margin-bottom: 2px;
}

#cog-btn {
  width: 40px !important;
  font-size: 16px;
  margin-bottom: 2px;
}

#import-btn,
#undo-btn {
  width: 40px !important;
  font-size: 16px;
}

.ace_editor_form {
  position: relative !important;
  height: 360px !important;
  width: 100% !important;
  border: 0 !important;
}

.ace_editor_modal {
  width: 100% !important;
  height: -webkit-fill-available !important;
}

#editor-container {
  display: flex;
  flex-direction: column;

  div.navbar-editor {
    .btn-link {
      outline: none;
    }
  }

  #editor {
    .ace_gutter {
      .ace_gutter-layer {
        margin-top: 5px !important;
      }

      .ace_gutter-active-line {
        margin-top: 5px;
      }
    }

    .ace_scroller {
      margin-top: 5px;
    }
  }

  #diff-table {
    height: 360px;

    .diff {
      border-bottom: 1px solid #e8e9ea;
    }
  }
}

#preview_error_toast {
  display: block;
  margin-bottom: 0;
}

#cog-popover {
  .cog-popover-dropdown {
    display: flex;
    align-items: center;

    .cog-popover-dropdown-title {
      width: 50px;
      margin-bottom: 5px;
      margin-right: 10px;
      font-weight: 600;
    }

    .dropdown {
      margin-left: auto;

      .dropdown-menu {
        min-width: 80px;
      }

      #mode-dropdown,
      #keybindings-dropdown,
      #themes-dropdown {
        width: 80px;
        text-align: left;

        .caret {
          position: absolute;
          right: 2px;
          top: 7px;
        }
      }
    }
  }
}
