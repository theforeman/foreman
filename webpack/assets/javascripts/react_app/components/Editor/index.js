import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './EditorActions';
import reducer from './EditorReducer';

import Editor from './Editor';

import {
  navFilteredHostsSelector,
  navHostsSelector,
  selectChosenHost,
  selectChosenView,
  selectDiffType,
  selectEditorName,
  selectErrorText,
  selectIsFetchingHosts,
  selectIsLoading,
  selectIsMasked,
  selectIsMaximized,
  selectIsReadOnly,
  selectIsRendering,
  selectIsSearchingHosts,
  selectIsSelectOpen,
  selectKeyBind,
  selectMode,
  selectPreviewResult,
  selectRenderedEditorValue,
  selectSearchQuery,
  selectShowError,
  selectTheme,
  selectValue,
} from './EditorSelectors';

// map state to props
const mapStateToProps = state => ({
  diffViewType: selectDiffType(state),
  editorName: selectEditorName(state),
  errorText: selectErrorText(state),
  filteredHosts: navFilteredHostsSelector(state),
  hosts: navHostsSelector(state),
  isFetchingHosts: selectIsFetchingHosts(state),
  isLoading: selectIsLoading(state),
  isMasked: selectIsMasked(state),
  isMaximized: selectIsMaximized(state),
  isRendering: selectIsRendering(state),
  isSearchingHosts: selectIsSearchingHosts(state),
  isSelectOpen: selectIsSelectOpen(state),
  keyBinding: selectKeyBind(state),
  mode: selectMode(state),
  previewResult: selectPreviewResult(state),
  renderedEditorValue: selectRenderedEditorValue(state),
  readOnly: selectIsReadOnly(state),
  searchQuery: selectSearchQuery(state),
  selectedHost: selectChosenHost(state),
  selectedView: selectChosenView(state),
  showError: selectShowError(state),
  theme: selectTheme(state),
  value: selectValue(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { editor: reducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(Editor);
