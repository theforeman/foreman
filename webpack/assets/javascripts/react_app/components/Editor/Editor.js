import React from 'react';
import PropTypes from 'prop-types';
import { ToastNotification } from 'patternfly-react';

import { noop } from '../../common/helpers';
import DiffView from '../DiffView/DiffView';
import EditorView from './components/EditorView';
import EditorNavbar from './components/EditorNavbar';
import EditorModal from './components/EditorModal';
import {
  EDITOR_THEMES,
  EDITOR_KEYBINDINGS,
  EDITOR_MODES,
} from './EditorConstants';
import './editor.scss';

class Editor extends React.Component {
  componentDidMount() {
    const {
      data: { hosts, templateClass, locked, template, type },
      initializeEditor,
      isMasked,
      isRendering,
      readOnly,
      previewResult,
      selectedView,
      showError,
    } = this.props;

    const initializeData = {
      hosts,
      isMasked,
      templateClass,
      isRendering,
      locked,
      readOnly,
      previewResult,
      selectedView,
      showError,
      template,
      type,
    };
    initializeEditor(initializeData);
  }

  render() {
    const {
      data: {
        name,
        renderPath,
        showHide,
        showImport,
        showPreview,
        showHostSelector,
        template,
        title,
      },
      changeDiffViewType,
      changeEditorValue,
      changeSetting,
      changeTab,
      diffViewType,
      dismissErrorToast,
      editorName,
      errorText,
      fetchAndPreview,
      filteredHosts,
      hosts,
      importFile,
      isFetchingHosts,
      isLoading,
      isMasked,
      isMaximized,
      isRendering,
      isSearchingHosts,
      isSelectOpen,
      keyBinding,
      mode,
      onHostSearch,
      onHostSelectToggle,
      onSearchClear,
      previewResult,
      previewTemplate,
      readOnly,
      renderedEditorValue,
      revertChanges,
      searchQuery,
      selectedHost,
      selectedView,
      showError,
      theme,
      toggleMaskValue,
      toggleModal,
      toggleRenderView,
      value,
    } = this.props;

    const editorViewProps = {
      value: isRendering ? previewResult : value,
      mode: isRendering ? 'Text' : mode,
      theme,
      keyBinding,
      onChange: isRendering ? noop : changeEditorValue,
      readOnly: readOnly || isRendering,
      isMasked,
    };
    const editorNameTab = {
      input: `${editorName}Code`,
      preview: `${editorName}Preview`,
    };

    return (
      <div id="editor-container">
        <ToastNotification
          id="preview_error_toast"
          type="error"
          className={showError ? '' : 'hidden'}
          onDismiss={() => dismissErrorToast()}
        >
          {errorText}
        </ToastNotification>
        <EditorNavbar
          changeDiffViewType={changeDiffViewType}
          changeTab={changeTab}
          changeSetting={changeSetting}
          modes={EDITOR_MODES}
          themes={EDITOR_THEMES}
          keyBindings={EDITOR_KEYBINDINGS}
          mode={isRendering ? 'Text' : mode}
          theme={theme}
          keyBinding={keyBinding}
          value={value}
          renderedEditorValue={renderedEditorValue}
          diffViewType={diffViewType}
          template={template}
          selectedView={selectedView}
          isDiff={template ? value !== template : false}
          isMasked={isMasked}
          isRendering={isRendering}
          isLoading={isLoading}
          isFetchingHosts={isFetchingHosts}
          isSearchingHosts={isSearchingHosts}
          importFile={importFile}
          showImport={showImport}
          showPreview={showPreview}
          showHostSelector={showHostSelector}
          showHide={showHide}
          revertChanges={revertChanges}
          previewTemplate={previewTemplate}
          hosts={hosts}
          filteredHosts={filteredHosts}
          selectedHost={selectedHost}
          renderPath={renderPath}
          toggleMaskValue={toggleMaskValue}
          toggleRenderView={toggleRenderView}
          toggleModal={toggleModal}
          previewResult={previewResult}
          searchQuery={searchQuery}
          onHostSelectToggle={onHostSelectToggle}
          onHostSearch={onHostSearch}
          onSearchClear={onSearchClear}
          isSelectOpen={isSelectOpen}
          showError={showError}
          fetchAndPreview={fetchAndPreview}
        />
        <EditorView
          {...editorViewProps}
          key="editorPreview"
          name={editorNameTab.preview}
          isSelected={selectedView === 'preview'}
          className="ace_editor_form ace_preview"
        />
        <EditorView
          {...editorViewProps}
          key="editorCode"
          name={editorNameTab.input}
          isSelected={selectedView === 'input'}
          className="ace_editor_form ace_input"
        />
        <div
          id="diff-table"
          className={selectedView === 'diff' ? '' : 'hidden'}
        >
          <DiffView
            oldText={template || ''}
            newText={value}
            viewType={diffViewType}
          />
        </div>
        <EditorModal
          key="editorModal"
          changeEditorValue={changeEditorValue}
          changeDiffViewType={changeDiffViewType}
          name={editorName}
          title={title}
          toggleModal={toggleModal}
          diffViewType={diffViewType}
          mode={mode}
          theme={theme}
          keyBinding={keyBinding}
          readOnly={readOnly}
          isMaximized={isMaximized}
          template={template || ''}
          editorValue={value}
          previewValue={previewResult}
          selectedView={selectedView}
          isMasked={isMasked}
          isRendering={isRendering}
        />
        {!readOnly && (
          <textarea className="hidden" name={name} value={value} readOnly />
        )}
      </div>
    );
  }
}

Editor.propTypes = {
  data: PropTypes.shape({
    showHide: PropTypes.bool,
    showImport: PropTypes.bool,
    showPreview: PropTypes.bool,
    showHostSelector: PropTypes.bool,
    template: PropTypes.string,
    templateClass: PropTypes.string,
    name: PropTypes.string,
    title: PropTypes.string,
    renderPath: PropTypes.string,
    hosts: PropTypes.array,
    locked: PropTypes.bool,
    type: PropTypes.string,
  }).isRequired,
  selectedHost: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    name: PropTypes.string,
  }).isRequired,
  changeDiffViewType: PropTypes.func.isRequired,
  changeEditorValue: PropTypes.func.isRequired,
  changeSetting: PropTypes.func.isRequired,
  changeTab: PropTypes.func.isRequired,
  diffViewType: PropTypes.string.isRequired,
  dismissErrorToast: PropTypes.func.isRequired,
  editorName: PropTypes.string.isRequired,
  errorText: PropTypes.string.isRequired,
  hosts: PropTypes.array.isRequired,
  filteredHosts: PropTypes.array.isRequired,
  importFile: PropTypes.func.isRequired,
  initializeEditor: PropTypes.func.isRequired,
  isMasked: PropTypes.bool.isRequired,
  isMaximized: PropTypes.bool.isRequired,
  isRendering: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isFetchingHosts: PropTypes.bool.isRequired,
  keyBinding: PropTypes.string.isRequired,
  mode: PropTypes.string.isRequired,
  previewTemplate: PropTypes.func.isRequired,
  readOnly: PropTypes.bool.isRequired,
  previewResult: PropTypes.string.isRequired,
  revertChanges: PropTypes.func.isRequired,
  selectedView: PropTypes.string.isRequired,
  showError: PropTypes.bool.isRequired,
  theme: PropTypes.string.isRequired,
  toggleMaskValue: PropTypes.func.isRequired,
  toggleModal: PropTypes.func.isRequired,
  toggleRenderView: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  renderedEditorValue: PropTypes.string.isRequired,
  isSelectOpen: PropTypes.bool.isRequired,
  searchQuery: PropTypes.string.isRequired,
  onHostSelectToggle: PropTypes.func.isRequired,
  onHostSearch: PropTypes.func.isRequired,
  onSearchClear: PropTypes.func.isRequired,
  isSearchingHosts: PropTypes.bool.isRequired,
  fetchAndPreview: PropTypes.func.isRequired,
};

export default Editor;
