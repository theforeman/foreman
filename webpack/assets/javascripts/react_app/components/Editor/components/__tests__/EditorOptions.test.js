import React from 'react';
import { mount } from '@theforeman/test';
import { testComponentSnapshotsWithFixtures } from '../../../../common/testHelpers';

import EditorOptions from '../EditorOptions';
import { editorOptions, showBooleans } from '../../Editor.fixtures';

const props = { ...editorOptions, ...showBooleans, isDiff: true };

const fixtures = {
  'renders EditorOptions': props,
};

describe('EditorOptions', () => {
  describe('EditorOptions', () =>
    testComponentSnapshotsWithFixtures(EditorOptions, fixtures));

  describe('simulate onClick', () => {
    const toggleMaskValue = jest.fn();
    const changeTab = jest.fn();
    const revertChanges = jest.fn();
    jest.mock('../EditorOptions');
    window.confirm = jest.fn(() => true);

    const diffWrapper = mount(
      <EditorOptions
        {...props}
        changeTab={changeTab}
        toggleMaskValue={toggleMaskValue}
        revertChanges={revertChanges}
        isDiff
        selectedView="diff"
      />
    );

    const inputWrapper = mount(
      <EditorOptions
        {...props}
        changeTab={changeTab}
        toggleMaskValue={toggleMaskValue}
        revertChanges={revertChanges}
        isDiff
      />
    );

    diffWrapper
      .find('#undo-btn')
      .at(0)
      .simulate('click');
    inputWrapper
      .find('#hide-btn')
      .at(0)
      .simulate('click');
    diffWrapper
      .find('#import-btn')
      .at(0)
      .simulate('click');

    expect(toggleMaskValue).toHaveBeenCalledTimes(1);
    expect(changeTab).toHaveBeenCalledTimes(1);
    expect(window.confirm).toHaveBeenCalledTimes(1);
  });
});
