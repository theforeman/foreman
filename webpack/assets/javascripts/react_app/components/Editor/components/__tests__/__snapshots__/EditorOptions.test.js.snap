// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorOptions EditorOptions renders EditorOptions 1`] = `
<div
  id="editor-dropdowns"
>
  <h4
    id="divider"
  >
    |
  </h4>
  <OverlayTrigger
    defaultOverlayShown={false}
    delayShow={500}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="mask-tooltip"
        placement="right"
      >
        Hide Content
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
      ]
    }
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="link"
      className="editor-button"
      disabled={false}
      id="hide-btn"
      onClick={[Function]}
    >
      <Icon
        name="eye-slash"
        size="lg"
        type="fa"
      />
    </Button>
  </OverlayTrigger>
  <OverlayTrigger
    defaultOverlayShown={false}
    delayShow={500}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="revert-tooltip"
        placement="right"
      >
        Revert Local Changes
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
      ]
    }
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="link"
      className="editor-button"
      disabled={false}
      id="undo-btn"
      onClick={[Function]}
    >
      <Icon
        name="restart"
        size="2x"
        type="pf"
      />
    </Button>
  </OverlayTrigger>
  <OverlayTrigger
    defaultOverlayShown={false}
    delayShow={500}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="import-tooltip"
        placement="right"
      >
        Import File
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
      ]
    }
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="link"
      className="import-button"
      disabled={false}
      id="import-btn"
      onClick={[Function]}
    >
      <Icon
        name="folder-open"
        size="lg"
        type="pf"
      />
      <FormControl
        bsClass="form-control"
        className="hidden"
        componentClass="input"
        inputRef={[Function]}
        onChange={[Function]}
        type="file"
      />
    </Button>
  </OverlayTrigger>
  <EditorSettings
    changeSetting={[Function]}
    keyBinding="Default"
    keyBindings={
      Array [
        "Default",
        "Emacs",
        "Vim",
      ]
    }
    mode="Ruby"
    modes={
      Array [
        "Text",
        "Json",
        "Ruby",
        "Html_ruby",
        "Sh",
        "Xml",
        "Yaml",
      ]
    }
    selectedView="input"
    theme="Monokai"
    themes={
      Array [
        "Github",
        "Monokai",
      ]
    }
  />
  <OverlayTrigger
    defaultOverlayShown={false}
    delayShow={500}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="fullscreen-tooltip"
        placement="right"
      >
        Maximize
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
      ]
    }
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="link"
      className="editor-button"
      disabled={false}
      id="fullscreen-btn"
      onClick={[Function]}
    >
      <Icon
        name="arrows-alt"
        size="lg"
        type="fa"
      />
    </Button>
  </OverlayTrigger>
</div>
`;
